text/0079-cdk-v2.md

```diff
@@ -0,0 +1,339 @@
+a relative path). This is problematic for the CDK case since CDK modules often
+expose types from other modules in their API (e.g. an `s3.Bucket` accepts a
+`kms.Key` as an option).
+
+To model this type of dependency, npm allows declaring a dependency as a ["peer
+dependency"](https://nodejs.org/es/blog/npm/peer-dependencies). This means
+consumers are expected to **directly** install all transitive dependencies. In
+the *aws-ecs* case, this means that instead of declaring a dependency only on
+*aws-ecs*, users would have to declare a dependency on all 23 transitive modules
+that *aws-ecs* depend on.
+
+This is an unacceptable user experience with our current tools.
+
+See the [appendix](#appendix-cdk-and-npm-dependencies) written by
+[@rix0rrr](https://github.com/rix0rrr) which describes this problem in detail.
+
+### Poor ergonomics due to too many modules
+
+In addition to the peer npm dependency issue described above, due to the fact
+that CDK is "hyper modular", in all languages, we have poor ergonomics when it
+comes to declaring dependencies. Since users are required to explicitly install
+a module for each service they use, even simple projects end up with dozens of
+direct CDK dependencies.
+
+## Experimental Modules
+
+In some cases, a CDK module can include both stable and experimental constructs.
+For example, a module like **@aws-cdk/aws-eks** includes stable CFN resources (L1s)
+but also a variety of experimental L2 classes.
+
+Our API stability guarantees and adherence to semantic versioning only applies
+to STABLE APIs and not to EXPERIMENTAL APIs.
+
+This situation is confusing to users. It is not uncommon for users to use
+experimental constructs without being aware and then be surprised by breaking
+changes in minor versions.
+
+The current approach makes it hard for us to specify the stability at the module
+level since the L1s are stable by definition, but all high-level constructs are
+still experimental.
+
+## Reusing constructs
+
+The AWS CDK demonstrates that leveraging general purpose programming languages
+to define complex configuration modules brings value to users. At the heart of
+the AWS CDK lies the "constructs programming model" which consists of a simple
+pattern that enables composition and synthesis.
+
+Recently, there were a few initiatives to build CDKs for other domains with
+similar characteristics. Examples are Kubernetes, Terraform and GitHub Actions.
+
+We wish to enable the creation of these "sister" CDK projects by extracting the
+constructs programming model into an independent library, which can be leveraged
+for any domain without taking a dependency on the AWS CDK.
+
+In order to enable composition of constructs across domains (e.g. Kubernetes app
+defined within an AWS CDK app), we will need to extract the `Construct` base
+class from the @aws-cdk/core module and into a separate module and AWS CDK users
+will need to take a dependency on this new module.
+
+# Design Summary
```

Comment:
I see another problem that I think a package split would address:

*Many people are mixing up L1 and L2 APIs* 

I see a lot of people using L2 classes, and then being disappointed when they can't plug them into L1 classes. This happens a lot in Python (where there's obviously no typechecker to tell you you're doing it wrong), but also happens in other languages where people call integration APIs between L2 classes to plug the result into an L1 (see recent changes to CloudWatch dashboards, mutliple people internally and externally were calling `toJson()` on `GraphWidgets` and plugging the result into a `CfnDashboard`, blissfully unaware that a special method needs to be called there to do a Token-aware JSON stringification).

Apparently our current communication methods of `Foo` and `CfnFoo` classes, plus notes that these are separate classes plus notes in every talk that we do aren't enough to dissuade people from doing this.

I would therefore like to propose we drop the L1-and-L2-in-a-single-package style of bundling, and publish 2 different packages, one with all L1s and one with all L2s.

Note that this needn't imply a `peerDependency` of L2-to-L1 library, for most reasonable use cases(*) a regular `dependency` of L2-to-L1 is sufficient (since by design we don't expose L1 types in the L2 API, and in fact jsii won't allow us to do that).

_(*) The existence of `construct.node.defaultChild` breaks this assumption a little, it's important that the consumer app and construct library agree on the version of the `CfnFoo` class that is accidentally exposed this way. Still, I expect this to be a niche use case that will mostly "just work"_

Reply:
Noted. I will incorporate this into the doc